src/awkward/_broadcasting.py

```diff
@@ -497,68 +498,100 @@ def broadcast_any_record():
+                                    dtype=np.int64,
                                 ),
-                                nplike=backend.index_nplike,
-                            )
+                                index_nplike.shape_item_as_index(dim_size),
+                            ),
+                            nplike=index_nplike,
+                        )
+                        break
 
-                nextinputs = []
-                for x in inputs:
-                    if isinstance(x, RegularArray):
-                        if dimsize > 1 and x.size == 1:
-                            nextinputs.append(
-                                x.content[: x.length * x.size]._carry(
-                                    tmpindex, allow_lazy=False
-                                )
-                            )
-                        elif x.size == dimsize:
-                            nextinputs.append(x.content[: x.length * x.size])
-                        elif dimsize == 0:
-                            nextinputs.append(x.content[:0])
-                        else:
-                            raise ValueError(
-                                "cannot broadcast RegularArray of size "
-                                "{} with RegularArray of size {} {}".format(
-                                    x.size, dimsize, in_function(options)
-                                )
+            # Here we have three possible broadcasting outcomes (by precedence):
+            # 1. any (exactly) size-0 content trims all other contents
+            # 2. any (exactly) size-1 content broadcasts to the common length
+            # 3. otherwise, recurse into the content as-is
+            nextinputs = []
+            for x in inputs:
+                if isinstance(x, RegularArray):
+                    x_size_known_to_be_one = (
+                        x.size is not unknown_length and x.size == 1
+                    )
+                    # If dimsize is known to be exactly zero, all contents are zero length
+                    if dimsize_known_to_be_zero:
+                        nextinputs.append(x.content[:0])
+                    # If we have a known size=1 content, then broadcast it to the dimension size
+                    elif dimsize_greater_than_one_if_known and x_size_known_to_be_one:
+                        nextinputs.append(
+                            x.content[: x.length * x.size]._carry(
+                                size_one_carry_index, allow_lazy=False
                             )
+                        )
+                    # Any unknown values or sizes are assumed to be correct as-is
+                    elif (
+                        dim_size is unknown_length
+                        or x.size is unknown_length
+                        or x.size == dim_size
+                    ):
+                        nextinputs.append(x.content[: x.length * x.size])
                     else:
```

Comment:
@jpivarski do you think that these rules are reasonable? It means that if we lose length information, we assume that neither the 1-length or 0-length broadcasting applies.

Reply:
If we don't know a RegularArray size at type-tracer time, we won't enter any specialized length-0 or length-1 logic, but we also don't know that the lengths are incompatible, so we'd return a type-tracer of unknown length—not raise an error.

At run-time (on a Dask worker), the length with either be length-0, length-1, or something else, and that something else will either be compatible or incompatible. The type-tracer time return value (an array with unknown length) is either compatible with the run-time result or the run-time result is an error. That's fine.

The only bad cases would be

  * if the type-tracer returned a result that is incompatible with the run-time result, such as a different number of dimensions (the type reported by the type-tracer is wrong), or
  * if the type-tracer raises an error that doesn't get raised at run-time (users would be able to do something in Awkward Array that they can't do in dask-awkward, and it would be hard for us to explain why).

As I understand it, what you have above is not either of these two bad cases.

Reply:
I think so. The form should be preserved, at least. What might be different is that if the `_carry` only happens on the non-typetracer branch, then the form of the result would differ (`ListArray` vs `ListOffsetArray`) if the content is a `ListOffsetArray`.